src/psyclone/tests/psyad/adjoint_visitor_test.py

```diff
@@ -413,6 +443,151 @@ def test_assignment_node(fortran_reader, fortran_writer):
+    # lower bound
+    adj_visitor = AdjointVisitor(["a", "b", "c", "lo"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The lower bound of a loop should not contain active variables, "
+            "but found 'lo'" in str(info.value))
+    # upper bound
+    adj_visitor = AdjointVisitor(["a", "b", "c", "hi"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The upper bound of a loop should not contain active variables, "
+            "but found 'hi'" in str(info.value))
+    # step
+    adj_visitor = AdjointVisitor(["a", "b", "c", "step"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The step of a loop should not contain active variables, "
+            "but found 'step'" in str(info.value))
+    # loop variable
+    adj_visitor = AdjointVisitor(["a", "b", "c", "i"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The loop iterator 'i' should not be an active variable."
+            in str(info.value))
+
+
+def test_loop_node_passive(fortran_reader):
+    '''Test that the loop_node method raises an exception if there are no
+    active variables within the supplied loop node. This is because
+    the schedule node should have already dealt with this case.
+
+    '''
+    tl_psyir = fortran_reader.psyir_from_source(TL_LOOP_CODE)
+    tl_loop = tl_psyir.walk(Loop)[0]
+    assert isinstance(tl_loop, Loop)
+    adj_visitor = AdjointVisitor(["d", "e"])
+    ad_psyir = adj_visitor._visit(tl_psyir)
+    ad_loop = ad_psyir.walk(Loop)[0]
+    assert isinstance(ad_loop, Loop)
+
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor.loop_node(tl_loop)
+    assert ("A passive loop node should not be processed by the loop_node() "
+            "method within the AdjointVisitor() class, as it should have been "
+            "dealt with by the schedule_node() method." in str(info.value))
+
+
+@pytest.mark.parametrize("in_bounds,out_bounds", [
+    ("lo,hi", "hi, lo, -1"),
+    ("lo,hi,1", "hi, lo, -1"),
+    ("lo,hi,-1", "hi - MOD(hi - lo, - 1), lo, 1"),
+    ("lo,hi,step", "hi - MOD(hi - lo, step), lo, -1 * step")])
+def test_loop_node_active(fortran_reader, fortran_writer, in_bounds,
+                          out_bounds):
+    '''Test that a loop_node containing active variables returns with its
+    loop order reversed and its loop body processed by the adjoint
+    visitor. Checks that appropriate offset code is generated when the
+    loop step is not, or might not be, 1.
+
+    '''
+    code = TL_LOOP_CODE.replace("lo,hi,step", in_bounds)
+    tl_psyir = fortran_reader.psyir_from_source(code)
+    tl_loop = tl_psyir.walk(Loop)[0]
+    assert isinstance(tl_loop, Loop)
+    adj_visitor = AdjointVisitor(["a", "b", "c"])
+    ad_psyir = adj_visitor(tl_psyir)
+    ad_loop = ad_psyir.walk(Loop)[0]
+    assert isinstance(ad_loop, Loop)
```

Comment:
You don't need this line, ad_loop has to be a Loop due to the previous line.

Reply:
Removed all places where this is done.